sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/Main.java

```diff
@@ -74,30 +78,97 @@
+    final BlockingQueue results;
+    NexmarkPerf perf = null;
+    Exception exception = null;
+
+    private Run(String[] args, NexmarkConfiguration configuration, BlockingQueue results) {
+      NexmarkOptions options = PipelineOptionsFactory.fromArgs(args).as(NexmarkOptions.class);
+      this.nexmarkLauncher = new NexmarkLauncher<>(options);
+      this.configuration = configuration;
+      this.results = results;
+    }
+
+    @Override
+    public void run() {
+      try {
+        try {
+          perf = nexmarkLauncher.run(configuration);
+        } catch (IOException | RuntimeException e) {
+          exception = e;
+        } finally {
+          results.put(this);
+        }
+      } catch (InterruptedException ie) {
+        return;
+      }
+    }
+
+    private NexmarkPerf get() throws IOException {
+      if (exception != null) {
+        if (exception instanceof IOException) {
+          throw (IOException) exception;
+        } else if (exception instanceof RuntimeException) {
+          throw (RuntimeException) exception;
+        }
+        // This should be unreachable.
+      }
+      assert perf != null;
+      return perf;
+    }
+  }
 
   /** Entry point. */
-  void runAll(OptionT options, NexmarkLauncher nexmarkLauncher) throws IOException {
+  void runAll(String[] args) throws IOException {
     Instant start = Instant.now();
+    NexmarkOptions options =
+        PipelineOptionsFactory.fromArgs(args).withValidation().as(NexmarkOptions.class);
+    ExecutorService executor = Executors.newFixedThreadPool(options.getNexmarkParallel());
+    BlockingQueue<Run> results = new LinkedBlockingQueue();
     Map<NexmarkConfiguration, NexmarkPerf> baseline = loadBaseline(options.getBaselineFilename());
     Map<NexmarkConfiguration, NexmarkPerf> actual = new LinkedHashMap<>();
     Iterable<NexmarkConfiguration> configurations = options.getSuite().getConfigurations(options);
 
     boolean successful = true;
     try {
-      // Run all the configurations.
+      // Schedule all the configurations.
       for (NexmarkConfiguration configuration : configurations) {
-        NexmarkPerf perf = nexmarkLauncher.run(configuration);
-        if (perf != null) {
-          if (perf.errors == null || perf.errors.size() > 0) {
-            successful = false;
+        executor.execute(new Run(args, configuration, results));
+        if (options.getNexmarkParallel() > 1) {
+          // Stagger start
+          try {
+            Thread.sleep(2000);
+          } catch (InterruptedException e) {
+            return;
           }
-          appendPerf(options.getPerfFilename(), configuration, perf);
-          actual.put(configuration, perf);
-          // Summarize what we've run so far.
-          saveSummary(null, configurations, actual, baseline, start, options);
         }
       }
+
+      // Collect all the results.
+      for (Object i : configurations) {
+        Run result;
+        try {
+          result = results.take();
```

Comment:
This looks like a wait to me. If I'm correct, please add timeout.

Reply:
Synced with Andrew. We have test timeout configured in test framework. Right now we rely on that one to kill the test if it hangs for too long.
We can add cleanup on timout logic in another PR.

Reply:
This is a wait, but there is no reasonable timeout that can be applied at this layer. This is waiting for a test to complete. These tests are variable runtime and are run for multiple hours in some places.